sentinelhub/sentinelhub_byoc.py

```diff
@@ -0,0 +1,264 @@
+
+from dataclasses_json import config as dataclass_config
+from dataclasses_json import dataclass_json, LetterCase, Undefined, CatchAll
+
+from .data_collections import DataCollection
+from .config import SHConfig
+from .constants import RequestType, MimeType
+from .download.sentinelhub_client import SentinelHubDownloadClient
+from .geometry import Geometry
+from .sh_utils import iter_pages, remove_undefined, from_sh_datetime, to_sh_datetime
+
+
+datetime_config = dataclass_config(encoder=to_sh_datetime, decoder=from_sh_datetime,
+                                   letter_case=LetterCase.CAMEL)
+
+geometry_config = dataclass_config(encoder=Geometry.get_geojson,
+                                   decoder=lambda x: Geometry.from_geojson(x) if x else None,
+                                   exclude=lambda x: x is None, letter_case=LetterCase.CAMEL)
+
+
+@dataclass_json(letter_case=LetterCase.CAMEL, undefined=Undefined.INCLUDE)
+@dataclass
+class ByocCollectionAdditionalData:
+    """ Dataclass to hold BYOC collection additional data
+    """
+    other_data: CatchAll
+    bands: Optional[dict] = None
+    max_meters_per_pixel: Optional[float] = None
+    max_meters_per_pixel_override: Optional[float] = None
+
+
+@dataclass_json(letter_case=LetterCase.CAMEL, undefined=Undefined.INCLUDE)
+@dataclass
+class ByocCollection:
+    """ Dataclass to hold BYOC collection data
+    """
+    name: str
+    s3_bucket: str
+    other_data: CatchAll
+    collection_id: Optional[str] = field(metadata=dataclass_config(field_name='id'), default=None)
+    user_id: Optional[str] = None
+    created: Optional[datetime] = field(metadata=datetime_config, default=None)
+    additional_data: Optional[ByocCollectionAdditionalData] = None
+    no_data: Optional[Union[int, float]] = None
+
+    def to_data_collection(self):
+        """ Returns a DataCollection enum for this BYOC collection
+        """
+        if self.additional_data and self.additional_data.bands:
+            band_names = tuple(self.additional_data.bands)
+            return DataCollection.define_byoc(collection_id=self.collection_id, bands=band_names)
+
+        return DataCollection.define_byoc(collection_id=self.collection_id)
+
+
+@dataclass_json(letter_case=LetterCase.CAMEL, undefined=Undefined.INCLUDE)
+@dataclass
+class ByocTile:
+    """ Dataclass to hold BYOC tile data
+    """
+    path: str
+    other_data: CatchAll
+    status: Optional[str] = None
+    tile_id: Optional[str] = field(metadata=dataclass_config(field_name='id'), default=None)
+    tile_geometry: Optional[Geometry] = field(metadata=geometry_config, default=None)
+    cover_geometry: Optional[Geometry] = field(metadata=geometry_config, default=None)
+    created: Optional[datetime] = field(metadata=datetime_config, default=None)
+    sensing_time: Optional[datetime] = field(metadata=datetime_config, default=None)
+    additional_data: Optional[dict] = None
+
+    @property
+    def tile_geometry_crs(self):
```

Comment:
I'm just wondering how useful properties `tile_geometry_crs`, `cover_geometry_crs`, and `geometry` are. Usually you either have a geometry for all tiles or you don't. And if a geometry is empty you might still want to get a CRS or geometry.

Feel free to ignore this comment, I'm also fine if we leave them as they are.

Reply:
I wanted the `_crs` because they can be different between the tiles. `_geometry` seemed useful to convert to `shapely` object, but in the end one can always call `tile_geometry.geometry` or `cover_geometry.geometry` to achieve the same. As to some extent "less is better than more", I would remove them.